core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -595,6 +596,9 @@ private Evaluator extractInclusiveDeleteExpression(ManifestReader reader) {
           if (bin.contains(cachedNewManifest) && bin.size() < minManifestsCountToMerge) {
             // not enough to merge, add all manifest files to the output list
             outputManifests.addAll(bin);
+          } else if ((!Collections.disjoint(bin, appendManifests)) && bin.size() < minManifestsCountToMerge) {
```

Comment:
I don't think the logic is quite right. The check above, `bin.contains(cachedNewManifest)` is intended to catch the last bin. Only the last bin is left unmerged, so that it can accumulate more manifests and isn't merged every time. But the bin before the last can be merged if it is full.

This logic would prevent any bin with an appended manifest from being merged, even if it isn't the last bin.

Maybe we should come up with a better way to detect the last bin, since the `cachedNewManifest` may not be present.

Reply:
thanks @rdblue for the explanation. Sorry to delay in reply !

> The check above, bin.contains(cachedNewManifest) is intended to catch the last bin. Only the last bin is left unmerged, so that it can accumulate more manifests and isn't merged every time. But the bin before the last can be merged if it is full.

it means if `bin.contains(cachedNewManifest)` is true, then this the last bin.
Because latest added manifests/files has to be in the last bin?

To handle the appendManifest case, we can maintain one more variables `cachedNewAppenedManifest`, which will be initialized by new `ManifestFile` supplied to `appendManifests(ManifestFile manifestFile)`

condition could be:
```
else if (bin.contains(cachedNewAppenedManifest) && bin.size() < minManifestsCountToMerge) {
            // not enough to merge, add all manifest files to the output list
            outputManifests.addAll(bin);
          }
```

which means, if the `cachedNewAppenedManifest` (latest appendedManifest file) is present in the bin, then it would be the last one.

Reply:
@rdblue please review at your convenience.